### ISSUE TITLE
UnicodeDecodeError on .mif headers

### ISSUE THREAD

**Issue by lrq3000** [MRtrix3/mrtrix3#1423]
I am using the latest MRTRIX3 version pulled from git one week ago.

When trying to preprocess with dwipreproc a .mif file just extracted from the DICOMs using mrcat, I got the following error:

```
dwipreproc: 
dwipreproc: Note that this script makes use of commands / algorithms that have relevant articles for citation; INCLUDING FROM EXTERNAL SOFTWARE PACKAGES. Please consult the help page (-help option) for more information.
dwipreproc: 
dwipreproc: Loading header for image file '/home/brain/neuro-csg-pipelines/dwi-scripts-allinone/proc/subject/DICOM/18080613/45090002/dwi.mif'
Traceback (most recent call last):
  File "/home/brain/neurotools/mrtrix3/bin/dwipreproc", line 74, in <module>
    image.check3DNonunity(path.fromUser(app.args.input, False))
  File "/home/brain/neurotools/mrtrix3/lib/mrtrix3/image.py", line 101, in check3DNonunity
    image_in = Header(image_in)
  File "/home/brain/neurotools/mrtrix3/lib/mrtrix3/image.py", line 20, in __init__
    data = json.load(f)
  File "/home/brain/anaconda2/lib/python2.7/json/__init__.py", line 291, in load
    **kw)
  File "/home/brain/anaconda2/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/home/brain/anaconda2/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/home/brain/anaconda2/lib/python2.7/json/decoder.py", line 380, in raw_decode
    obj, end = self.scan_once(s, idx)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xe9 in position 11: invalid continuation byte
```

The .mif file header was the following:

```
mrtrix image
dim: 128,128,66,167
vox: 2,2,2.60001,nan
layout: -0,-1,+2,+3
datatype: UInt16LE
transform: 0.9975141308963952, -0.07045276546044839, -0.001401007684390283, -116.7135868207305
transform: 0.07043740924297978, 0.9974755109864508, -0.009009996612834239, -149.7998817382754
transform: 0.002032250266677158, 0.00888891564219859, 0.9999584277051247, -131.0549761408544
FlipAngle: 90
RepetitionTime: 3.3
command_history: mrcat "." "." "." "dwi.mif"  (version=3.0_RC3-62-gd6656921)
comments: Subject (000000A) [MR] ep2d_diff_mddw_30_p2_s3_b700
comments: study: IRM étude fonctionnelle de l'encéphale [ ORIGINAL PRIMARY DIFFUSION NONE ]
comments: DOB: XX/XX/XX
comments: DOS: XX/XX/2018 XX:XX:XX
dw_scheme: 0,0,0,0
```

The header line that was causing the issue was this one:
```comments: study: IRM étude fonctionnelle de l'encéphale [ ORIGINAL PRIMARY DIFFUSION NONE ]```

After renaming to remove the accentuated characters, the preprocessing went fine!

This is probably related to #1296 and #953. It seems the patches didn't help here (this bug was also reproduced on another computer with another subject from the same study).

If I can suggest something, maybe such an error could be caught and in this case the data filtered to remove all accentuated characters? This could be a failsafe in case the decoding didn't properly work.

**Comment by Lestropie** [comment]
Hi Stephen,

Thanks for the exceptionally clear report. This appears to be more of what was reported in #1347.

What we need to decide on is the solution:

- While the error is arising within the Python code, it's going to be slightly awkward to *solve* the issue within Python: The error is coming from deep inside the Python JSON module, in a way that I can't easily catch and handle (e.g. unlike the solution [here](https://github.com/MRtrix3/mrtrix3/commit/724e91b71b95d185d1e6113909d5a623250b5613), where we just accept that some characters may not appear correctly, in your case it's deep within a library function, so once the exception is raised the JSON file import can't proceed).
What would need to happen is that the `image.Header.__init__()` function would try to read the JSON file based on UTF-8, and if it fails, create a new JSON file where the contents of the file have been copied using `decode('utf-8', errors='replace')`, and then load that new file using the `json` module.

- The more general solution is for our DICOM C++ import code to identify when DICOM fields are stored using non-standard encoding, and explicitly either convert any strings into UTF-8 or replace any non-ASCII characters with their closest equivalents.
This is a path that @jdtournier and I have been having a silent stand-off as to who takes responsibility for implementing... Unlike a language like Python, which has plenty of useful libraries for text handling, C++ has very little in the way of exotic text support.

If you're able to provide an example problematic data set, that would be useful.

Rob

**Comment by thijsdhollander** [comment]
Coincidentally (even though not strictly about this): http://community.mrtrix.org/t/population-template-error-population-template-line-442-typeerror-decode-takes-no-keyword-arguments/1829

**Comment by Lestropie** [comment]
On second thought, the quick solution may be pretty easy... @lrq3000 Feel like giving branch `python_image_header_decode` a try?

**Comment by lrq3000** [comment]
@Lestropie Sorry for the delay, didn't see your replies before, I tried the branch you mentioned and unfortunately there is a new error:

```
dwipreproc: Loading header for image file '/xxx/dwi.mif'
Traceback (most recent call last):
  File "/home/brain/neurotools/mrtrix3/bin/dwipreproc", line 74, in <module>
    image.check3DNonunity(path.fromUser(app.args.input, False))
  File "/home/brain/neurotools/mrtrix3/lib/mrtrix3/image.py", line 105, in check3DNonunity
    image_in = Header(image_in)
  File "/home/brain/neurotools/mrtrix3/lib/mrtrix3/image.py", line 24, in __init__
    data = json.load(f.read().decode('utf-8', errors='replace'))
  File "/home/brain/anaconda2/lib/python2.7/json/__init__.py", line 287, in load
    return loads(fp.read(),
AttributeError: 'unicode' object has no attribute 'read'
```

I will send a copy of the dwi.mif so you guys can work on it :-)

**Comment by lrq3000** [comment]
@Lestropie Just saw your message, thank you very much for the patch, there was just a typo that I fixed in #1465. I tested the patch (with the addendum in my PR) and it works perfectly fine!
